internal/signalfx-agent/pkg/monitors/prometheusexporter/conversion.go

```diff
@@ -24,66 +22,110 @@ func counterExtractor(m *dto.Metric) float64 {
 	return m.GetCounter().GetValue()
 }
 
-func convertMetricFamily(mf *dto.MetricFamily) []*datapoint.Datapoint {
+func convertMetricFamily(sm pmetric.ScopeMetrics, mf *dto.MetricFamily, applyDimensions func(attrs pcommon.Map)) {
 	if mf.Type == nil || mf.Name == nil {
-		return nil
+		return
 	}
 	switch *mf.Type {
 	case dto.MetricType_GAUGE:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.GaugeF, gaugeExtractor)
+		makeGaugeDataPoints(sm, *mf.Name, mf.Metric, gaugeExtractor, applyDimensions)
 	case dto.MetricType_COUNTER:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.CumulativeF, counterExtractor)
+		makeSumDataPoints(sm, *mf.Name, mf.Metric, counterExtractor, applyDimensions)
 	case dto.MetricType_UNTYPED:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.GaugeF, untypedExtractor)
+		makeGaugeDataPoints(sm, *mf.Name, mf.Metric, untypedExtractor, applyDimensions)
 	case dto.MetricType_SUMMARY:
-		return makeSummaryDatapoints(*mf.Name, mf.Metric)
+		makeSummaryDatapoints(sm, *mf.Name, mf.Metric, applyDimensions)
 	// TODO: figure out how to best convert histograms, in particular the
 	// upper bound value
 	case dto.MetricType_HISTOGRAM:
-		return makeHistogramDatapoints(*mf.Name, mf.Metric)
+		makeHistogramDatapoints(sm, *mf.Name, mf.Metric, applyDimensions)
 	default:
-		return nil
 	}
 }
 
-func makeSimpleDatapoints(name string, ms []*dto.Metric, dpf dpFactory, e extractor) []*datapoint.Datapoint {
```

Comment:
Does it make more sense to have flagged, additive changes so that rolling back can occur in the same version (or this is opt-in). Moving everything in one go can set users up for poor experience.

Reply:
Yes, it does. A feature gate would make the most sense. I can do this, I didn't think this changeset would get so involved to be honest.